Gems/Atom/RHI/Vulkan/Code/Source/RHI/Scope.cpp

```diff
@@ -649,11 +649,32 @@ namespace AZ
         void Scope::OptimizeBarrier(const Barrier& unoptimizedBarrier, BarrierSlot slot)
         {
             uint32_t slotIndex = aznumeric_caster(slot);
-            AZStd::vector<Barrier>& barriers =
-                CanOptimizeBarrier(unoptimizedBarrier, slot) ? m_globalBarriers[slotIndex] : m_scopeBarriers[slotIndex];
+            bool canOptimize = CanOptimizeBarrier(unoptimizedBarrier, slot);
             auto filteredBarrier = unoptimizedBarrier;
             filteredBarrier.m_dstStageMask = RHI::FilterBits(unoptimizedBarrier.m_dstStageMask, m_deviceSupportedPipelineStageFlags);
-            filteredBarrier.m_srcStageMask = RHI::FilterBits(unoptimizedBarrier.m_srcStageMask, m_deviceSupportedPipelineStageFlags);            
+            filteredBarrier.m_srcStageMask = RHI::FilterBits(unoptimizedBarrier.m_srcStageMask, m_deviceSupportedPipelineStageFlags);
+
+            // a render pass can only automatically convert image layouts if a multi-aspect image has the same or compatible layouts for all
+            // aspects. This workaround covers the case for depth/stencil images where one aspect is in a transfer state:
+            if (canOptimize && filteredBarrier.m_type == VK_STRUCTURE_TYPE_IMAGE_MEMORY_BARRIER && UsesRenderpass() &&
+                IsRenderAttachmentUsage(filteredBarrier.m_attachment->GetUsage()) &&
+                (filteredBarrier.m_imageBarrier.subresourceRange.aspectMask == VK_IMAGE_ASPECT_DEPTH_BIT ||
+                 filteredBarrier.m_imageBarrier.subresourceRange.aspectMask == VK_IMAGE_ASPECT_STENCIL_BIT) &&
+                (filteredBarrier.m_imageBarrier.oldLayout == VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL ||
+                 filteredBarrier.m_imageBarrier.oldLayout == VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL))
+            {
+                auto oldLayout = filteredBarrier.m_imageBarrier.oldLayout;
+
+                // insert the newLayout as oldLayout into m_globalBarriers as the RenderPass uses oldLayout to determine the initial layout
+                filteredBarrier.m_imageBarrier.oldLayout = filteredBarrier.m_imageBarrier.newLayout;
+                OptimizeBarrier(filteredBarrier, slot);
+
+                // insert an actual barrier for transfer -> newLayout transition into m_scopeBarriers
+                filteredBarrier.m_imageBarrier.oldLayout = oldLayout;
+                canOptimize = false;
+            }
+
```

Comment:
Let me copy the relevant part of the PR description:

> The change in Scope.cpp is a workaround for a specific issue we encountered where for a multi-aspect (depth and stencil) image with differing layouts for the aspects the layout transition is not done correctly. The situation is as follows:
> 
> 1. We copy a depth stencil image (final layout: transfer dest)
> 2. We run a compute pass that reads only the depth image (final layout: depth read only optimal for depth, while stencil is still in transfer dest)
> 3. We run a render pass where we use it as actual depth stencil attachment
> 
> The Vulkan render pass needs a working initial layout and that is determined by the global barriers which are not actually submitted, since the render pass does the necessary layout transition. It can only do so if the layout of the aspect masks is the same or at least can be combined to a valid one (eg. depth read only and stencil read only to depth stencil read only). However with the two layouts the image has at the time it wants to combine them and the combination doesn't work because the stencil aspect is still in transfer dest and not in a attachment optimal or read only optimal layout. Therefore it is necessary to insert a barrier that transitions the stencil aspect to one of these which in our workaround is done by directly going to the target layout that the render pass would transition to. The combination then works because the (recursive) OptimizeBarrier call in the if inserts the necessary global barrier that the render pass uses to determine the initial layout combination. We then set `canOptimize` to false to actually insert a barrier in the scope barriers that gets executed for the actual transition from transfer dest to the target layout. Of course this is a very specific use case and there are likely other instances where something like this can happen for depth/stencil images (or other currently unsupported multi-aspect images), e.g., the other way around where depth is still in a transfer state and stencil is correct. The way barriers and layout transitions are currently handled doesn't allow for a nice solution here really unfortunately.

Reply:
Your solution here doesn't exactly work unfortunately since the first of the barriers (which is usually stencil, since prologue barriers are traversed in reverse) is already optimized away, i.e., it doesn't end up in the `scopeBarriers` and thus the second cannot check overlaps between it and the first one.

The alternative I have working is to check before the `find_if` whether (1) the usage is `DepthStencil`, (2) the number of aspects in the aspect mask is only 1, (3) the old layout differs from the new one and (4) the image actually has 2 aspects (depth and stencil). If all these conditions apply, I return false, i.e., don't optimize and this works.

The only "issue" with this solution and the one you proposed is, that the condition also applies to other renderpasses, where the renderpass could handle the layout transition, because the layouts can be combined. E.g., depth is attachment optimal, stencil is read only -> can be combined to `VK_IMAGE_LAYOUT_DEPTH_ATTACHMENT_STENCIL_READ_ONLY_OPTIMAL` for the renderpass's initial layout. However, we would now do the transition manually with a barrier. That might or might not be an issue, because it could impact performance, but we couldn't find any information about this online.

The only solution to avoid this would be to check all unoptimized_barriers and check if the layouts can be combined, because the information of both barriers is need to decide for both of them whether they can be optimized or not.